From 676ce898fd730c4edf7e326951f928690a68fc99 Mon Sep 17 00:00:00 2001
From: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
Date: Tue, 26 Dec 2023 21:41:25 +0530
Subject: [PATCH] mac80211: Update link conf bssid indicator with actual bss number

The mbssid cnt is the count of total number of mbssid elements instead of
max BSSID indicator.
Therefore, we fix this by reading the Max BSSID Indicator field directly.

Signed-off-by: Ramasamy Kaliappan <quic_rkaliapp@quicinc.com>
---
 net/mac80211/cfg.c         |  6 ++++--
 net/mac80211/ieee80211_i.h | 14 ++++++++++++++
 2 files changed, 18 insertions(+), 2 deletions(-)

--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1358,7 +1358,7 @@ ieee80211_assign_beacon(struct ieee80211
 		pos = ieee80211_copy_rnr_beacon(pos, new->rnr_ies, rnr);
 		/* update bssid_indicator */
 		link_conf->bssid_indicator =
-			ilog2(__roundup_pow_of_two(mbssid->cnt + 1));
+			ieee80211_get_max_bssid_indicator(new->mbssid_ies);
 	}
 
 	if (csa) {
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -1211,6 +1211,19 @@ ieee80211_vif_get_shift(struct ieee80211
 	return shift;
 }
 
+static inline u8
+ieee80211_get_max_bssid_indicator(struct cfg80211_mbssid_elems *mbssid_elems)
+{
+	u8 *bssid_indicator_offset;
+
+	if (!mbssid_elems || !mbssid_elems->cnt)
+		return 0;
+
+	bssid_indicator_offset = mbssid_elems->elem[0].data + 2;
+
+	return *bssid_indicator_offset;
+}
+
 static inline int
 ieee80211_get_mbssid_beacon_len(struct cfg80211_mbssid_elems *elems,
 				struct cfg80211_rnr_elems *rnr_elems,
